lib/datastore/index.js

```diff
@@ -1,132 +1,260 @@
 
   /**
-   * Encode a doc (an object), to a block (a Buffer)
-   * @param {Doc} doc
-   * @returns {Block}
+   * Create a new DataType in the DataStore
+   * @param {Object} options
+   * @param {String} options.name
+   * @param {Object} options.schema
+   * @param {BlockPrefix} options.blockPrefix
+   * @param {String} options.extraColumns
+   * @returns
    */
-  encode(doc) {
-    return this.#dataType.encode(doc)
+  dataType(options) {
+    if (this.#dataTypes.has(options.name)) {
+      return this.#dataTypes.get(options.name)
+    }
+
+    const { name, schema, extraColumns, blockPrefix } = options
+
+    const dataType = new DataType({
+      name,
+      identityPublicKey: this.#identityPublicKey,
+      keyPair: this.#keyPair,
+      schema,
+      blockPrefix,
+      corestore: this.#corestore,
+      sqlite: this.#sqlite,
+      extraColumns,
+    })
+
+    this.#dataTypes.set(name, dataType)
+
+    for (const core of dataType.cores) {
+      core.on('ready', () => {
+        this.#indexer.addCore(dataType.core)
+      })
+    }
+
+    return dataType
   }
 
   /**
-   * Decode a block (a Buffer), to a doc (an object)
    * @param {Block} block
-   * @returns {Doc}
+   * @returns {DataType}
    */
-  decode(block) {
-    return this.#dataType.decode(block)
+  getDataTypeForBlock(block) {
+    return this.dataTypes.find((dataType) => {
+      return dataType.isType(block)
+    })
   }
 
   /**
-   * Get a doc by id
-   * @param {string} id
-   * @returns {Doc}
+   * @param {string} dataTypeName
+   * @param {Doc} data
+   * @returns {Promise<Doc>}
+   * @throws {Error}
    */
-  getById(id) {
-    return this.#indexer.query(`id = '${id}'`)[0]
+  async create(dataTypeName, data) {
+    if (!this.#dataTypes.has(dataTypeName)) {
+      throw new Error(`Data type ${dataTypeName} not found`)
+    }
+
+    const dataType = this.#dataTypes.get(dataTypeName)
+    return dataType.create(data)
   }
 
   /**
-   * Create a doc
+   * @param {string} dataTypeName
    * @param {Doc} data
    * @returns {Promise<Doc>}
+   * @throws {Error}
    */
-  async create(data) {
-    const doc = Object.assign(data, {
-      id: data.id || randomBytes(8).toString('hex'),
-      version: `${this.#writer.key.toString('hex')}@${this.#writer.length}`,
-      created_at: new Date().toISOString(),
-    })
-
-    if (!doc.links) {
-      doc.links = []
+  async update(dataTypeName, data) {
+    if (!this.#dataTypes.has(dataTypeName)) {
+      throw new Error(`Data type ${dataTypeName} not found`)
     }
 
-    this.validate(doc)
-    const encodedDoc = this.encode(doc)
-    await this.#writer.append(encodedDoc)
+    const dataType = this.#dataTypes.get(dataTypeName)
+    return dataType.update(data)
+  }
 
-    const indexing = new Promise((resolve) => {
-      this.#indexer.onceWriteDoc(doc.version, (doc) => {
-        resolve(doc)
-      })
+  /**
+   * Wait for the indexer to finish available blocks
+   * @returns {Promise<void>}
+   */
+  async indexing() {
```

Comment:
We need to be very careful about how this might be used outside a testing context (ok to use it as a helper in tests).
In real usage a device is going to be switching between "indexing" and "idle" multiple times — imagine sync over a slow network where new blocks are arriving more slowly than blocks are indexed. You might want to consider "indexing is done when nothing new is arriving for XX seconds". 
For now this is used only in the test helper, but I want to make sure we don't leave it exposed and start using it in other code without considering the limitations/fagility of it.